### ISSUE TITLE
Make CircuitBreaker delay time mockable for testing

### ISSUE THREAD

**Issue by dugsmith** [failsafe-lib/failsafe#357]
I'm using `CircuitBreaker` in its [Standalone Usage](https://failsafe.dev/circuit-breaker/#standalone-usage). My `CircuitBreaker` is configured `withFailureRateThreshold(50, 2, Duration.ofSeconds(30))` to open when 50% of requests over a 30 second period fail. 

I'd like to be able to unit test around the `CircuitBreaker` behavior so that I know that the closed / open / half-open behaviors are correct.

The challenge appears to be that `OpenState` uses `System.nanoTime()` to calculate the delay. This is very difficult to mock. I started going down the road of using AspectJ, but this is a mixed Kotlin/Java Android project, and the [best library](https://github.com/Ibotta/gradle-aspectj-pipeline-plugin) I found to implement AspectJ is no longer maintained because Android Gradle Plugin 8.0 will no longer support bytecode transformation. 

If you switched to use something like `java.time.Instant.now()`, that would make test mocking much easier.

I see that in your own tests you just configure very short delay times and use `Thread.sleep()`. That's okay, but it's not nearly as flexible as being able to mock time itself.

**Comment by jhalterman** [comment]
Yes, I can imagine mocking `System.nanoTime` would be difficult, though I would think the same is true with`Instant.now` since they're both static - or am I missing something? From what I understand, the problem you're looking to avoid, using System time for unit tests, can only be avoided if we have some overridable mechanism for supplying the current time. Maybe you can give me a better idea of what you were thinking.

**Comment by Tembrel** [comment]
Something like the [Clock abstraction](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html) might be useful, but it would mean some refactoring, because I don't think that it works at nanosecond granularity. (Mind you, I don't think one needs or wants true nanosecond granularity in cases where Failsafe is used, but the code currently is written that way.)

**Comment by dugsmith** [comment]
Thank you for the quick response, @jhalterman. `System.nanoTime()` is a native method, and MockK doesn't mock native methods. `Instant.now()` is static but not native, so it is easily mockable by libraries like MockK.

If you did want to provide a Clock abstraction for unit testing, that would be even better. But a quicker fix might be to just use `java.time.Instant` and document how to mock it.

**Comment by jhalterman** [comment]
I'm not sure I'd document mocking `OpenState` since it's an internal, package private class that isn't really meant to be used directly. Can you give me a better idea of how you're using it and testing against it?
